documentation/developing/modding_examples/blindfold_assist.mdx

```diff
@@ -0,0 +1,187 @@
+- aligning the character against walls or other level geometry
+- being able to move in straight lines and/or consistent distances
+- manipulating the camera in consistent patterns
+
+When it comes to Jak, the game's responsive engine provides maybe too much flexibility in controlling Jak's movement and the camera 😅. This ends up hindering the consistency of blindfolded setups - so for this mod we'll actually be limiting some functionality! Here are the two changes I set out to tackle (thanks Borrdi for the suggestions):
+- limit analog stick input to 8 directions ↑ ↗ → ↘ ↓ ↙ ← ↖ (every 45 degrees)
+- add a button combo to (attempt to) center the camera behind Jak
+
+I'll go through things in more detail below, but you can [click here to see the full diff](https://github.com/OpenGOAL-Unofficial-Mods/blindfold-assist/commit/1cd217498b5efa501c4980241d253a6b09e8c330) of changes for this mod. It's only 69 lines across 3 files, and most of that is duplicated code for the left vs right analog sticks.
+
+<ReactPlayer controls url={require("./video/blindfold-assist.mp4").default} className="blog-video"/>
+
+## Limiting analog stick input to 8 directions
+
+### Where/how is controller input defined?
+
+On the OpenGOAL side, the controller input is represented by the [`cpad-info` type](https://github.com/open-goal/jak-project/blob/ee3bc182e13a440f90e20a6f7be7d3b294a7e56a/goal_src/jak1/engine/ps2/pad.gc#L49-L100). In particular, the below fields (some from the parent `hw-cpad` type) relate to the analog stick inputs:
+
+```lisp
+;; data that comes directly from hardware. it's 32 bytes + type tag (ignored in C kernel).
+(deftype hw-cpad (basic)
+  (
+    ...
+    (rightx  uint8     :offset-assert 8) ;; right stick xdir
+    (righty  uint8     :offset-assert 9) ;; right stick ydir
+    (leftx   uint8     :offset-assert 10) ;; left stick xdir
+    (lefty   uint8     :offset-assert 11) ;; left stick ydir
+    ...
+    )
+  )
+
+;; data from hardware + additional info calculated here.
+(deftype cpad-info (hw-cpad)
+  (
+    ...
+    (stick0-dir         float           :offset-assert 72)
+    (stick0-speed       float           :offset-assert 76)
+    ...
+    )
+  )
+  ```
+  
+You can sort of think of `rightx` and `righty` as the X and Y "coordinates" for the position of the right analog stick, read directly from the hardware. Since they are `uint8`s they range in value from 0 (left/up) to 255 (right/down) - so both would be 128 while neutral (assuming no stick drift). The same applies for `leftx` and `lefty` for the left analog stick, however there is an extra step where `stick0-dir` and `stick0-speed` are computed from the values of `leftx` and `lefty`, happening once per frame [within `service-cpads`](https://github.com/open-goal/jak-project/blob/ee3bc182e13a440f90e20a6f7be7d3b294a7e56a/goal_src/jak1/engine/ps2/pad.gc#L344-L349).
```

Comment:
`127` is actually the neutral value for the analogs

Reply:
is it? https://github.com/open-goal/jak-project/blob/ee3bc182e13a440f90e20a6f7be7d3b294a7e56a/goal_src/jak1/engine/ps2/pad.gc#L355-L359
```opengoal
   ;; analog is invalid? set to zero.
   (set! (-> pad leftx) (the-as uint 128))
   (set! (-> pad lefty) (the-as uint 128))
   (set! (-> pad rightx) (the-as uint 128))
   (set! (-> pad righty) (the-as uint 128))
```

Reply:
hah, well maybe from GOAL's perspective it's not.  From the hardware/emulator itself 127 is the true center.  Though realistically both are going to be in the deadzone so it doesn't really matter.

https://github.com/open-goal/jak-project/blob/ee3bc182e13a440f90e20a6f7be7d3b294a7e56a/game/system/newpad.cpp#L239-L246

Reply:
lol well wait does that mean it introduces a slight up/left bias outside the deadzone though? (or is it down/right? too tired to think)

Reply:
I suppose so yeah, though definitely imperceptible.

Reply:
127.5 is the actual "true center." But it can't be represented as an int, obviously. The actual math for analog stick values to movement stats is done here: https://github.com/open-goal/jak-project/blob/master/goal_src/jak1/engine/ps2/pad.gc#L344-L345 it returns -1.0 if the stick is all the way to the left and 0.9921875 if it's all the way to the right, but it's 0.9921875 if the stick is all the way up and -1.0 if it's all the way down.

So, either 128 or 127 are fine.